In [1]:
import pandas as pd

In [2]:
data = pd.read_excel("MO14-Round-1-Dealing-With-Data-Workbook (1).xlsx", sheet_name="Usage", header=None)
data

0
0         3 PM  Mon 24th-Mar-2014___0.384 kwh  
1                5AM  15-Aug-2014___1.201  kwh 
2          __8PM Thu 20-Mar-2014____1.523 kwh  
3               6PM  23rd-Apr-2014___0.424  kwh
4     _1AM  Friday 19th-Dec-2014___0.209  kwh  
...                                         ...
8755         _1AM  Fri 07th-Nov-2014_0.084 kwh 
8756                _6AM 20-May-2014__1.027 kwh
8757   __2 AM  Tuesday 8th-Apr-2014___0.052 kwh
8758            9 PM 27th-Jan-2014_1.428  kwh  
8759          _12 PM 25th-Aug-2014_0.528  kwh  

[8760 rows x 1 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       8760 non-null   object
dtypes: object(1)
memory usage: 68.6+ KB


In [4]:
data.isnull().sum()

0    0
dtype: int64

In [5]:
data.rename(columns={0:"column1"}, inplace=True)
data.head()

column1
0      3 PM  Mon 24th-Mar-2014___0.384 kwh  
1             5AM  15-Aug-2014___1.201  kwh 
2       __8PM Thu 20-Mar-2014____1.523 kwh  
3            6PM  23rd-Apr-2014___0.424  kwh
4  _1AM  Friday 19th-Dec-2014___0.209  kwh

In [6]:
data.column1 = data.column1.apply(lambda x: x.strip(" _"))
data

column1
0        3 PM  Mon 24th-Mar-2014___0.384 kwh
1              5AM  15-Aug-2014___1.201  kwh
2           8PM Thu 20-Mar-2014____1.523 kwh
3            6PM  23rd-Apr-2014___0.424  kwh
4     1AM  Friday 19th-Dec-2014___0.209  kwh
...                                      ...
8755        1AM  Fri 07th-Nov-2014_0.084 kwh
8756              6AM 20-May-2014__1.027 kwh
8757  2 AM  Tuesday 8th-Apr-2014___0.052 kwh
8758           9 PM 27th-Jan-2014_1.428  kwh
8759          12 PM 25th-Aug-2014_0.528  kwh

[8760 rows x 1 columns]

### KWH

In [7]:
data["column1"].str.rsplit("_", n=1)

0          [3 PM  Mon 24th-Mar-2014__, 0.384 kwh]
1                [5AM  15-Aug-2014__, 1.201  kwh]
2             [8PM Thu 20-Mar-2014___, 1.523 kwh]
3              [6PM  23rd-Apr-2014__, 0.424  kwh]
4       [1AM  Friday 19th-Dec-2014__, 0.209  kwh]
                          ...                    
8755          [1AM  Fri 07th-Nov-2014, 0.084 kwh]
8756                [6AM 20-May-2014_, 1.027 kwh]
8757    [2 AM  Tuesday 8th-Apr-2014__, 0.052 kwh]
8758             [9 PM 27th-Jan-2014, 1.428  kwh]
8759            [12 PM 25th-Aug-2014, 0.528  kwh]
Name: column1, Length: 8760, dtype: object

In [8]:
data["kwh"] = data["column1"].str.rsplit("_", n=1).str[1]
data.head()

column1         kwh
0     3 PM  Mon 24th-Mar-2014___0.384 kwh   0.384 kwh
1           5AM  15-Aug-2014___1.201  kwh  1.201  kwh
2        8PM Thu 20-Mar-2014____1.523 kwh   1.523 kwh
3         6PM  23rd-Apr-2014___0.424  kwh  0.424  kwh
4  1AM  Friday 19th-Dec-2014___0.209  kwh  0.209  kwh

In [9]:
data["kwh"] = data["kwh"].str.rstrip(" kwh ")
data.head()

column1    kwh
0     3 PM  Mon 24th-Mar-2014___0.384 kwh  0.384
1           5AM  15-Aug-2014___1.201  kwh  1.201
2        8PM Thu 20-Mar-2014____1.523 kwh  1.523
3         6PM  23rd-Apr-2014___0.424  kwh  0.424
4  1AM  Friday 19th-Dec-2014___0.209  kwh  0.209

### Split date

In [10]:
data.column1.str.rstrip(" kwh").str.rsplit("_")

0          [3 PM  Mon 24th-Mar-2014, , , 0.384]
1                 [5AM  15-Aug-2014, , , 1.201]
2            [8PM Thu 20-Mar-2014, , , , 1.523]
3               [6PM  23rd-Apr-2014, , , 0.424]
4        [1AM  Friday 19th-Dec-2014, , , 0.209]
                         ...                   
8755            [1AM  Fri 07th-Nov-2014, 0.084]
8756                 [6AM 20-May-2014, , 1.027]
8757    [2 AM  Tuesday 8th-Apr-2014, , , 0.052]
8758                [9 PM 27th-Jan-2014, 1.428]
8759               [12 PM 25th-Aug-2014, 0.528]
Name: column1, Length: 8760, dtype: object

In [11]:
data["date"] = data.column1.str.rstrip(" kwh").str.rsplit("_").str[0]
data.head()

column1    kwh                       date
0     3 PM  Mon 24th-Mar-2014___0.384 kwh  0.384    3 PM  Mon 24th-Mar-2014
1           5AM  15-Aug-2014___1.201  kwh  1.201           5AM  15-Aug-2014
2        8PM Thu 20-Mar-2014____1.523 kwh  1.523        8PM Thu 20-Mar-2014
3         6PM  23rd-Apr-2014___0.424  kwh  0.424         6PM  23rd-Apr-2014
4  1AM  Friday 19th-Dec-2014___0.209  kwh  0.209  1AM  Friday 19th-Dec-2014

In [12]:
data = data.drop("column1", axis = 1)
data.head()

kwh                       date
0  0.384    3 PM  Mon 24th-Mar-2014
1  1.201           5AM  15-Aug-2014
2  1.523        8PM Thu 20-Mar-2014
3  0.424         6PM  23rd-Apr-2014
4  0.209  1AM  Friday 19th-Dec-2014

### Day

In [13]:
data.date.str.rsplit(" ",n= 1)

0          [3 PM  Mon, 24th-Mar-2014]
1                 [5AM , 15-Aug-2014]
2              [8PM Thu, 20-Mar-2014]
3               [6PM , 23rd-Apr-2014]
4        [1AM  Friday, 19th-Dec-2014]
                    ...              
8755        [1AM  Fri, 07th-Nov-2014]
8756               [6AM, 20-May-2014]
8757    [2 AM  Tuesday, 8th-Apr-2014]
8758            [9 PM, 27th-Jan-2014]
8759           [12 PM, 25th-Aug-2014]
Name: date, Length: 8760, dtype: object

In [14]:
data["day"] = data.date.str.rsplit(" ",n= 1).str[1]
data.head()

kwh                       date            day
0  0.384    3 PM  Mon 24th-Mar-2014  24th-Mar-2014
1  1.201           5AM  15-Aug-2014    15-Aug-2014
2  1.523        8PM Thu 20-Mar-2014    20-Mar-2014
3  0.424         6PM  23rd-Apr-2014  23rd-Apr-2014
4  0.209  1AM  Friday 19th-Dec-2014  19th-Dec-2014

In [15]:
data["day"].unique()[:50]

array(['24th-Mar-2014', '15-Aug-2014', '20-Mar-2014', '23rd-Apr-2014',
       '19th-Dec-2014', '19th-Aug-2014', '7th-Jul-2014', '25-Nov-2014',
       '14-Aug-2014', '25th-Jan-2014', '11th-Feb-2014', '11-Jul-2014',
       '28-Dec-2014', '01-Apr-2014', '04th-Feb-2014', '11th-May-2014',
       '24-Jun-2014', '19th-Jun-2014', '23rd-Sep-2014', '26-Jul-2014',
       '23rd-Jun-2014', '01-Jul-2014', '23-Dec-2014', '20-Dec-2014',
       '17th-May-2014', '28th-Apr-2014', '8th-Mar-2014', '21-Jan-2014',
       '17-Jan-2014', '2-Apr-2014', '1st-Jun-2014', '24th-Oct-2014',
       '03-Oct-2014', '15-Nov-2014', '08-May-2014', '25-Apr-2014',
       '8-Sep-2014', '03rd-Sep-2014', '24-Aug-2014', '23-Mar-2014',
       '9-Jan-2014', '15th-May-2014', '31-Oct-2014', '15th-Aug-2014',
       '15-Mar-2014', '21st-Jan-2014', '26th-Oct-2014', '02nd-May-2014',
       '30-Nov-2014', '13th-Apr-2014'], dtype=object)

In [16]:
data["day"] = data["day"].str.replace("th","")
data["day"] = data["day"].str.replace("rd","")
data["day"] = data["day"].str.replace("st","")
data["day"] = data["day"].str.replace("nd","")

In [17]:
data.head(10)

kwh                       date          day
0  0.384    3 PM  Mon 24th-Mar-2014  24-Mar-2014
1  1.201           5AM  15-Aug-2014  15-Aug-2014
2  1.523        8PM Thu 20-Mar-2014  20-Mar-2014
3  0.424         6PM  23rd-Apr-2014  23-Apr-2014
4  0.209  1AM  Friday 19th-Dec-2014  19-Dec-2014
5  1.228     5AM  Tue 19th-Aug-2014  19-Aug-2014
6  1.296     12PM  Mon 7th-Jul-2014   7-Jul-2014
7  0.437      7 AM  Tue 25-Nov-2014  25-Nov-2014
8  0.523           8AM  14-Aug-2014  14-Aug-2014
9  2.052         4PM  25th-Jan-2014  25-Jan-2014

### Hour and Weekday

In [18]:
data.date.str.rsplit(" ",n = 1)

0          [3 PM  Mon, 24th-Mar-2014]
1                 [5AM , 15-Aug-2014]
2              [8PM Thu, 20-Mar-2014]
3               [6PM , 23rd-Apr-2014]
4        [1AM  Friday, 19th-Dec-2014]
                    ...              
8755        [1AM  Fri, 07th-Nov-2014]
8756               [6AM, 20-May-2014]
8757    [2 AM  Tuesday, 8th-Apr-2014]
8758            [9 PM, 27th-Jan-2014]
8759           [12 PM, 25th-Aug-2014]
Name: date, Length: 8760, dtype: object

In [19]:
data["hour_weekday"] = data.date.str.rsplit(" ",n = 1).str[0]
data.head()

kwh                       date          day hour_weekday
0  0.384    3 PM  Mon 24th-Mar-2014  24-Mar-2014    3 PM  Mon
1  1.201           5AM  15-Aug-2014  15-Aug-2014         5AM 
2  1.523        8PM Thu 20-Mar-2014  20-Mar-2014      8PM Thu
3  0.424         6PM  23rd-Apr-2014  23-Apr-2014         6PM 
4  0.209  1AM  Friday 19th-Dec-2014  19-Dec-2014  1AM  Friday

In [20]:
data = data.drop("date", axis = 1)
data.head()

kwh          day hour_weekday
0  0.384  24-Mar-2014    3 PM  Mon
1  1.201  15-Aug-2014         5AM 
2  1.523  20-Mar-2014      8PM Thu
3  0.424  23-Apr-2014         6PM 
4  0.209  19-Dec-2014  1AM  Friday

### Hour

In [21]:
data["hour_weekday"] = data["hour_weekday"].str.replace(" ","")
data

kwh          day hour_weekday
0     0.384  24-Mar-2014       3PMMon
1     1.201  15-Aug-2014          5AM
2     1.523  20-Mar-2014       8PMThu
3     0.424  23-Apr-2014          6PM
4     0.209  19-Dec-2014    1AMFriday
...     ...          ...          ...
8755  0.084  07-Nov-2014       1AMFri
8756  1.027  20-May-2014          6AM
8757  0.052   8-Apr-2014   2AMTuesday
8758  1.428  27-Jan-2014          9PM
8759  0.528  25-Aug-2014         12PM

[8760 rows x 3 columns]

In [22]:
data["hour"] = data["hour_weekday"].str[:4]
data

kwh          day hour_weekday  hour
0     0.384  24-Mar-2014       3PMMon  3PMM
1     1.201  15-Aug-2014          5AM   5AM
2     1.523  20-Mar-2014       8PMThu  8PMT
3     0.424  23-Apr-2014          6PM   6PM
4     0.209  19-Dec-2014    1AMFriday  1AMF
...     ...          ...          ...   ...
8755  0.084  07-Nov-2014       1AMFri  1AMF
8756  1.027  20-May-2014          6AM   6AM
8757  0.052   8-Apr-2014   2AMTuesday  2AMT
8758  1.428  27-Jan-2014          9PM   9PM
8759  0.528  25-Aug-2014         12PM  12PM

[8760 rows x 4 columns]

In [23]:
data["hour"].str.rstrip("MTWFS")

0        3P
1        5A
2        8P
3        6P
4        1A
       ... 
8755     1A
8756     6A
8757     2A
8758     9P
8759    12P
Name: hour, Length: 8760, dtype: object

In [24]:
data["hour"] = data["hour"].str.rstrip("MTWFS") + "M"
data.head()

kwh          day hour_weekday hour
0  0.384  24-Mar-2014       3PMMon  3PM
1  1.201  15-Aug-2014          5AM  5AM
2  1.523  20-Mar-2014       8PMThu  8PM
3  0.424  23-Apr-2014          6PM  6PM
4  0.209  19-Dec-2014    1AMFriday  1AM

In [25]:
data["hour"].unique()

array(['3PM', '5AM', '8PM', '6PM', '1AM', '12PM', '7AM', '8AM', '4PM',
       '12AM', '3AM', '11PM', '7PM', '11AM', '9AM', '1PM', '10PM', '10AM',
       '6AM', '2AM', '4AM', '2PM', '9PM', '5PM'], dtype=object)

### Weekday

In [26]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   kwh           8760 non-null   object
 1   day           8760 non-null   object
 2   hour_weekday  8760 non-null   object
 3   hour          8760 non-null   object
dtypes: object(4)
memory usage: 273.9+ KB


In [27]:
data["day"] = data["day"].astype("datetime64")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   kwh           8760 non-null   object        
 1   day           8760 non-null   datetime64[ns]
 2   hour_weekday  8760 non-null   object        
 3   hour          8760 non-null   object        
dtypes: datetime64[ns](1), object(3)
memory usage: 273.9+ KB


In [28]:
data["day_name"] = data["day"].dt.day_name()

In [29]:
data.head()

kwh        day hour_weekday hour   day_name
0  0.384 2014-03-24       3PMMon  3PM     Monday
1  1.201 2014-08-15          5AM  5AM     Friday
2  1.523 2014-03-20       8PMThu  8PM   Thursday
3  0.424 2014-04-23          6PM  6PM  Wednesday
4  0.209 2014-12-19    1AMFriday  1AM     Friday

In [30]:
data.drop("hour_weekday", axis = 1, inplace = True)
data.head()

kwh        day hour   day_name
0  0.384 2014-03-24  3PM     Monday
1  1.201 2014-08-15  5AM     Friday
2  1.523 2014-03-20  8PM   Thursday
3  0.424 2014-04-23  6PM  Wednesday
4  0.209 2014-12-19  1AM     Friday

# Question 1

In [31]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   kwh       8760 non-null   object        
 1   day       8760 non-null   datetime64[ns]
 2   hour      8760 non-null   object        
 3   day_name  8760 non-null   object        
dtypes: datetime64[ns](1), object(3)
memory usage: 273.9+ KB


In [32]:
data["kwh"] = data["kwh"].astype("float64")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   kwh       8760 non-null   float64       
 1   day       8760 non-null   datetime64[ns]
 2   hour      8760 non-null   object        
 3   day_name  8760 non-null   object        
dtypes: datetime64[ns](1), float64(1), object(2)
memory usage: 273.9+ KB


In [33]:
round(data["kwh"].mean(),3)

0.782

# Question 2

In [34]:
data.head()

kwh        day hour   day_name
0  0.384 2014-03-24  3PM     Monday
1  1.201 2014-08-15  5AM     Friday
2  1.523 2014-03-20  8PM   Thursday
3  0.424 2014-04-23  6PM  Wednesday
4  0.209 2014-12-19  1AM     Friday

In [35]:
data["Month"] = data["day"].dt.month_name()
data.head()

kwh        day hour   day_name     Month
0  0.384 2014-03-24  3PM     Monday     March
1  1.201 2014-08-15  5AM     Friday    August
2  1.523 2014-03-20  8PM   Thursday     March
3  0.424 2014-04-23  6PM  Wednesday     April
4  0.209 2014-12-19  1AM     Friday  December

In [36]:
data.groupby("Month")["kwh"].mean()["February"].round(3)

0.833

# Question 3

In [37]:
data.groupby("day_name")["kwh"].mean().nlargest(1)

day_name
Sunday    0.898157
Name: kwh, dtype: float64

# Question 4

In [38]:
data.head()

kwh        day hour   day_name     Month
0  0.384 2014-03-24  3PM     Monday     March
1  1.201 2014-08-15  5AM     Friday    August
2  1.523 2014-03-20  8PM   Thursday     March
3  0.424 2014-04-23  6PM  Wednesday     April
4  0.209 2014-12-19  1AM     Friday  December

In [39]:
data["hour"].str.rstrip("PM")

0        3
1       5A
2        8
3        6
4       1A
        ..
8755    1A
8756    6A
8757    2A
8758     9
8759    12
Name: hour, Length: 8760, dtype: object

In [40]:
data["new_hour"] = data["hour"].str.rstrip("PM").transform(lambda x: x if x.startswith("12")
                                                           else
                                                           int(x) + 12 
                                                           if "A" not in x                                                        
                                                           else x.replace("A", ""))
data 

kwh        day  hour   day_name     Month new_hour
0     0.384 2014-03-24   3PM     Monday     March       15
1     1.201 2014-08-15   5AM     Friday    August        5
2     1.523 2014-03-20   8PM   Thursday     March       20
3     0.424 2014-04-23   6PM  Wednesday     April       18
4     0.209 2014-12-19   1AM     Friday  December        1
...     ...        ...   ...        ...       ...      ...
8755  0.084 2014-11-07   1AM     Friday  November        1
8756  1.027 2014-05-20   6AM    Tuesday       May        6
8757  0.052 2014-04-08   2AM    Tuesday     April        2
8758  1.428 2014-01-27   9PM     Monday   January       21
8759  0.528 2014-08-25  12PM     Monday    August       12

[8760 rows x 6 columns]

In [41]:
data[data["hour"] == "12AM"]

kwh        day  hour   day_name     Month new_hour
12    0.845 2014-12-28  12AM     Sunday  December      12A
29    0.546 2014-01-17  12AM     Friday   January      12A
30    0.100 2014-04-02  12AM  Wednesday     April      12A
31    1.670 2014-06-01  12AM     Sunday      June      12A
41    0.540 2014-01-09  12AM   Thursday   January      12A
...     ...        ...   ...        ...       ...      ...
8704  0.389 2014-08-27  12AM  Wednesday    August      12A
8720  0.924 2014-12-21  12AM     Sunday  December      12A
8733  1.904 2014-01-05  12AM     Sunday   January      12A
8734  1.123 2014-08-09  12AM   Saturday    August      12A
8748  0.994 2014-12-13  12AM   Saturday  December      12A

[365 rows x 6 columns]

In [42]:
data[data["hour"] == "12AM"] = data[data["hour"] == "12AM"].transform(lambda x: x.replace("12A", 0))

In [43]:
data.head(20)

kwh        day  hour   day_name      Month new_hour
0   0.384 2014-03-24   3PM     Monday      March       15
1   1.201 2014-08-15   5AM     Friday     August        5
2   1.523 2014-03-20   8PM   Thursday      March       20
3   0.424 2014-04-23   6PM  Wednesday      April       18
4   0.209 2014-12-19   1AM     Friday   December        1
5   1.228 2014-08-19   5AM    Tuesday     August        5
6   1.296 2014-07-07  12PM     Monday       July       12
7   0.437 2014-11-25   7AM    Tuesday   November        7
8   0.523 2014-08-14   8AM   Thursday     August        8
9   2.052 2014-01-25   4PM   Saturday    January       16
10  0.509 2014-02-11   4PM    Tuesday   February       16
11  0.547 2014-07-11   1AM     Friday       July        1
12  0.845 2014-12-28  12AM     Sunday   December        0
13  0.557 2014-04-01   8PM    Tuesday      April       20
14  0.283 2014-02-04   3AM    Tuesday   February        3
15  0.344 2014-05-11  11PM     Sunday        May       23
16  2.948 2014-06-24   7PM    Tuesday       June       19
17  0.378 2014-06-19   1AM   Thursday       June        1
18  0.963 2014-09-23   8PM    Tuesday  September       20
19  1.469 2014-07-26  12PM   Saturday       July       12

In [44]:
data["new_hour"] = data["new_hour"].transform(lambda x: str(x) + ":00:00")
data.head(13)

kwh        day  hour   day_name     Month  new_hour
0   0.384 2014-03-24   3PM     Monday     March  15:00:00
1   1.201 2014-08-15   5AM     Friday    August   5:00:00
2   1.523 2014-03-20   8PM   Thursday     March  20:00:00
3   0.424 2014-04-23   6PM  Wednesday     April  18:00:00
4   0.209 2014-12-19   1AM     Friday  December   1:00:00
5   1.228 2014-08-19   5AM    Tuesday    August   5:00:00
6   1.296 2014-07-07  12PM     Monday      July  12:00:00
7   0.437 2014-11-25   7AM    Tuesday  November   7:00:00
8   0.523 2014-08-14   8AM   Thursday    August   8:00:00
9   2.052 2014-01-25   4PM   Saturday   January  16:00:00
10  0.509 2014-02-11   4PM    Tuesday  February  16:00:00
11  0.547 2014-07-11   1AM     Friday      July   1:00:00
12  0.845 2014-12-28  12AM     Sunday  December   0:00:00

In [45]:
data["new_hour"].nunique()

24

In [46]:
data.head()

kwh        day hour   day_name     Month  new_hour
0  0.384 2014-03-24  3PM     Monday     March  15:00:00
1  1.201 2014-08-15  5AM     Friday    August   5:00:00
2  1.523 2014-03-20  8PM   Thursday     March  20:00:00
3  0.424 2014-04-23  6PM  Wednesday     April  18:00:00
4  0.209 2014-12-19  1AM     Friday  December   1:00:00

In [47]:
data["day"] = data["day"].astype("str")

In [48]:
def concat(x):
    return str(x["day"]) + " " + str(x["new_hour"])

In [49]:
data["time"] = data.apply(concat, axis = 1)

In [50]:
data

kwh         day  hour   day_name     Month  new_hour  \
0     0.384  2014-03-24   3PM     Monday     March  15:00:00   
1     1.201  2014-08-15   5AM     Friday    August   5:00:00   
2     1.523  2014-03-20   8PM   Thursday     March  20:00:00   
3     0.424  2014-04-23   6PM  Wednesday     April  18:00:00   
4     0.209  2014-12-19   1AM     Friday  December   1:00:00   
...     ...         ...   ...        ...       ...       ...   
8755  0.084  2014-11-07   1AM     Friday  November   1:00:00   
8756  1.027  2014-05-20   6AM    Tuesday       May   6:00:00   
8757  0.052  2014-04-08   2AM    Tuesday     April   2:00:00   
8758  1.428  2014-01-27   9PM     Monday   January  21:00:00   
8759  0.528  2014-08-25  12PM     Monday    August  12:00:00   

                     time  
0     2014-03-24 15:00:00  
1      2014-08-15 5:00:00  
2     2014-03-20 20:00:00  
3     2014-04-23 18:00:00  
4      2014-12-19 1:00:00  
...                   ...  
8755   2014-11-07 1:00:00  
8756   2014-05-20 6:00:00  
8757   2014-04-08 2:00:00  
8758  2014-01-27 21:00:00  
8759  2014-08-25 12:00:00  

[8760 rows x 7 columns]

In [51]:
data["time"] = data["time"].astype("datetime64")

In [52]:
data.head()

kwh         day hour   day_name     Month  new_hour                time
0  0.384  2014-03-24  3PM     Monday     March  15:00:00 2014-03-24 15:00:00
1  1.201  2014-08-15  5AM     Friday    August   5:00:00 2014-08-15 05:00:00
2  1.523  2014-03-20  8PM   Thursday     March  20:00:00 2014-03-20 20:00:00
3  0.424  2014-04-23  6PM  Wednesday     April  18:00:00 2014-04-23 18:00:00
4  0.209  2014-12-19  1AM     Friday  December   1:00:00 2014-12-19 01:00:00

In [53]:
data.resample('4H', on='time').kwh.sum()

time
2014-01-01 00:00:00     1.752
2014-01-01 04:00:00     6.576
2014-01-01 08:00:00     2.948
2014-01-01 12:00:00     3.176
2014-01-01 16:00:00    10.588
                        ...  
2014-12-31 04:00:00     3.731
2014-12-31 08:00:00     1.666
2014-12-31 12:00:00     1.896
2014-12-31 16:00:00     5.372
2014-12-31 20:00:00     3.900
Freq: 4H, Name: kwh, Length: 2190, dtype: float64

In [54]:
data.resample('4H', on='time').kwh.sum().max()

14.386000000000001

# Question 5

In [55]:
data.head()

kwh         day hour   day_name     Month  new_hour                time
0  0.384  2014-03-24  3PM     Monday     March  15:00:00 2014-03-24 15:00:00
1  1.201  2014-08-15  5AM     Friday    August   5:00:00 2014-08-15 05:00:00
2  1.523  2014-03-20  8PM   Thursday     March  20:00:00 2014-03-20 20:00:00
3  0.424  2014-04-23  6PM  Wednesday     April  18:00:00 2014-04-23 18:00:00
4  0.209  2014-12-19  1AM     Friday  December   1:00:00 2014-12-19 01:00:00

In [56]:
data["kwh"].describe().T

count    8760.000000
mean        0.781878
std         0.759824
min         0.048000
25%         0.282000
50%         0.523000
75%         0.994000
max         5.141000
Name: kwh, dtype: float64

In [57]:
data.groupby("Month")["kwh"].sum()

Month
April         151.242
August        633.390
December      568.739
February      559.719
January      1042.660
July         1255.524
June          904.247
March         420.748
May           417.157
November      232.420
October       313.719
September     349.686
Name: kwh, dtype: float64

In [58]:
cost_data = data.groupby("Month")["kwh"].sum().reset_index()
cost_data

Month       kwh
0       April   151.242
1      August   633.390
2    December   568.739
3    February   559.719
4     January  1042.660
5        July  1255.524
6        June   904.247
7       March   420.748
8         May   417.157
9    November   232.420
10    October   313.719
11  September   349.686

In [59]:
contract = pd.read_excel("MO14-Round-1-Dealing-With-Data-Workbook (1).xlsx").copy()
contract

Unnamed: 0 Unnamed: 1 Unnamed: 2  Unnamed: 3 Unnamed: 4    Unnamed: 5  \
0         NaN        NaN        NaN         NaN        NaN           NaN   
1         NaN        NaN        NaN         NaN        NaN           NaN   
2         NaN        NaN        NaN         NaN        NaN           NaN   
3         NaN        NaN        NaN         NaN        NaN           NaN   
4         NaN        NaN        NaN         NaN        NaN           NaN   
5         NaN        NaN        NaN         NaN        NaN           NaN   
6         NaN        NaN        NaN         NaN        NaN           NaN   
7         NaN    No Flex        NaN         NaN        NaN  Monthly Flex   
8   Flat Rate       0.21      $/kWh         NaN    January           0.2   
9         NaN        NaN        NaN         NaN   February          0.19   
10        NaN        NaN        NaN         NaN      March          0.17   
11        NaN        NaN        NaN         NaN      April          0.18   
12        NaN        NaN        NaN         NaN        May          0.22   
13        NaN        NaN        NaN         NaN       June          0.27   
14        NaN        NaN        NaN         NaN       July          0.24   
15        NaN        NaN        NaN         NaN     August          0.19   
16        NaN        NaN        NaN         NaN  September          0.18   
17        NaN        NaN        NaN         NaN    October          0.15   
18        NaN        NaN        NaN         NaN   November          0.14   
19        NaN        NaN        NaN         NaN   December          0.19   
20        NaN        NaN        NaN         NaN        NaN           NaN   
21        NaN        NaN        NaN         NaN        NaN           NaN   
22        NaN        NaN        NaN         NaN        NaN           NaN   
23        NaN        NaN        NaN         NaN        NaN           NaN   
24        NaN        NaN        NaN         NaN        NaN           NaN   
25        NaN        NaN        NaN         NaN        NaN           NaN   
26        NaN        NaN        NaN         NaN        NaN           NaN   
27        NaN        NaN        NaN         NaN        NaN           NaN   
28        NaN        NaN        NaN         NaN        NaN           NaN   
29        NaN        NaN        NaN         NaN        NaN           NaN   
30        NaN        NaN        NaN         NaN        NaN           NaN   
31        NaN        NaN        NaN         NaN        NaN           NaN   

   Unnamed: 6  Unnamed: 7       Unnamed: 8   Unnamed: 9 Unnamed: 10  
0         NaN         NaN              NaN          NaN         NaN  
1         NaN         NaN              NaN          NaN         NaN  
2         NaN         NaN              NaN          NaN         NaN  
3         NaN         NaN              NaN          NaN         NaN  
4         NaN         NaN              NaN          NaN         NaN  
5         NaN         NaN              NaN          NaN         NaN  
6         NaN         NaN              NaN          NaN         NaN  
7         NaN         NaN              NaN  Hourly Flex         NaN  
8       $/kWh         NaN   12AM until 1AM          0.1       $/kWh  
9       $/kWh         NaN    1AM until 2AM          0.1       $/kWh  
10      $/kWh         NaN    2AM until 3AM          0.1       $/kWh  
11      $/kWh         NaN    3AM until 4AM          0.1       $/kWh  
12      $/kWh         NaN    4AM until 5AM         0.15       $/kWh  
13      $/kWh         NaN    5AM until 6AM          0.2       $/kWh  
14      $/kWh         NaN    6AM until 7AM         0.24       $/kWh  
15      $/kWh         NaN    7AM until 8AM         0.24       $/kWh  
16      $/kWh         NaN    8AM until 9AM         0.26       $/kWh  
17      $/kWh         NaN   9AM until 10AM          0.2       $/kWh  
18      $/kWh         NaN  10AM until 11AM          0.2       $/kWh  
19      $/kWh         NaN  11AM until 12PM         0.26       $/kWh  
20        NaN         Na

In [60]:
monthly_data = contract.iloc[8:20,4:6]
monthly_data

Unnamed: 4 Unnamed: 5
8     January        0.2
9    February       0.19
10      March       0.17
11      April       0.18
12        May       0.22
13       June       0.27
14       July       0.24
15     August       0.19
16  September       0.18
17    October       0.15
18   November       0.14
19   December       0.19

In [61]:
monthly_data.rename(columns={"Unnamed: 4" : "Month", "Unnamed: 5" : "Flex"}, inplace = True)
monthly_data

Month  Flex
8     January   0.2
9    February  0.19
10      March  0.17
11      April  0.18
12        May  0.22
13       June  0.27
14       July  0.24
15     August  0.19
16  September  0.18
17    October  0.15
18   November  0.14
19   December  0.19

In [62]:
cost_data.sort_values("Month")

Month       kwh
0       April   151.242
1      August   633.390
2    December   568.739
3    February   559.719
4     January  1042.660
5        July  1255.524
6        June   904.247
7       March   420.748
8         May   417.157
9    November   232.420
10    October   313.719
11  September   349.686

In [63]:
amount_data = pd.merge(monthly_data, cost_data, on = "Month")
amount_data

Month  Flex       kwh
0     January   0.2  1042.660
1    February  0.19   559.719
2       March  0.17   420.748
3       April  0.18   151.242
4         May  0.22   417.157
5        June  0.27   904.247
6        July  0.24  1255.524
7      August  0.19   633.390
8   September  0.18   349.686
9     October  0.15   313.719
10   November  0.14   232.420
11   December  0.19   568.739

In [64]:
amount_data["cost"] = amount_data["Flex"] * amount_data["kwh"]
amount_data

Month  Flex       kwh     cost
0     January   0.2  1042.660  208.532
1    February  0.19   559.719  106.347
2       March  0.17   420.748  71.5272
3       April  0.18   151.242  27.2236
4         May  0.22   417.157  91.7745
5        June  0.27   904.247  244.147
6        July  0.24  1255.524  301.326
7      August  0.19   633.390  120.344
8   September  0.18   349.686  62.9435
9     October  0.15   313.719  47.0579
10   November  0.14   232.420  32.5388
11   December  0.19   568.739   108.06

In [65]:
round(amount_data["cost"].sum(), 2)

1421.82

# Question 6

In [66]:
contract

Unnamed: 0 Unnamed: 1 Unnamed: 2  Unnamed: 3 Unnamed: 4    Unnamed: 5  \
0         NaN        NaN        NaN         NaN        NaN           NaN   
1         NaN        NaN        NaN         NaN        NaN           NaN   
2         NaN        NaN        NaN         NaN        NaN           NaN   
3         NaN        NaN        NaN         NaN        NaN           NaN   
4         NaN        NaN        NaN         NaN        NaN           NaN   
5         NaN        NaN        NaN         NaN        NaN           NaN   
6         NaN        NaN        NaN         NaN        NaN           NaN   
7         NaN    No Flex        NaN         NaN        NaN  Monthly Flex   
8   Flat Rate       0.21      $/kWh         NaN    January           0.2   
9         NaN        NaN        NaN         NaN   February          0.19   
10        NaN        NaN        NaN         NaN      March          0.17   
11        NaN        NaN        NaN         NaN      April          0.18   
12        NaN        NaN        NaN         NaN        May          0.22   
13        NaN        NaN        NaN         NaN       June          0.27   
14        NaN        NaN        NaN         NaN       July          0.24   
15        NaN        NaN        NaN         NaN     August          0.19   
16        NaN        NaN        NaN         NaN  September          0.18   
17        NaN        NaN        NaN         NaN    October          0.15   
18        NaN        NaN        NaN         NaN   November          0.14   
19        NaN        NaN        NaN         NaN   December          0.19   
20        NaN        NaN        NaN         NaN        NaN           NaN   
21        NaN        NaN        NaN         NaN        NaN           NaN   
22        NaN        NaN        NaN         NaN        NaN           NaN   
23        NaN        NaN        NaN         NaN        NaN           NaN   
24        NaN        NaN        NaN         NaN        NaN           NaN   
25        NaN        NaN        NaN         NaN        NaN           NaN   
26        NaN        NaN        NaN         NaN        NaN           NaN   
27        NaN        NaN        NaN         NaN        NaN           NaN   
28        NaN        NaN        NaN         NaN        NaN           NaN   
29        NaN        NaN        NaN         NaN        NaN           NaN   
30        NaN        NaN        NaN         NaN        NaN           NaN   
31        NaN        NaN        NaN         NaN        NaN           NaN   

   Unnamed: 6  Unnamed: 7       Unnamed: 8   Unnamed: 9 Unnamed: 10  
0         NaN         NaN              NaN          NaN         NaN  
1         NaN         NaN              NaN          NaN         NaN  
2         NaN         NaN              NaN          NaN         NaN  
3         NaN         NaN              NaN          NaN         NaN  
4         NaN         NaN              NaN          NaN         NaN  
5         NaN         NaN              NaN          NaN         NaN  
6         NaN         NaN              NaN          NaN         NaN  
7         NaN         NaN              NaN  Hourly Flex         NaN  
8       $/kWh         NaN   12AM until 1AM          0.1       $/kWh  
9       $/kWh         NaN    1AM until 2AM          0.1       $/kWh  
10      $/kWh         NaN    2AM until 3AM          0.1       $/kWh  
11      $/kWh         NaN    3AM until 4AM          0.1       $/kWh  
12      $/kWh         NaN    4AM until 5AM         0.15       $/kWh  
13      $/kWh         NaN    5AM until 6AM          0.2       $/kWh  
14      $/kWh         NaN    6AM until 7AM         0.24       $/kWh  
15      $/kWh         NaN    7AM until 8AM         0.24       $/kWh  
16      $/kWh         NaN    8AM until 9AM         0.26       $/kWh  
17      $/kWh         NaN   9AM until 10AM          0.2       $/kWh  
18      $/kWh         NaN  10AM until 11AM          0.2       $/kWh  
19      $/kWh         NaN  11AM until 12PM         0.26       $/kWh  
20        NaN         Na

In [67]:
no_flex = contract.iloc[8,[0,1]]
no_flex

Unnamed: 0    Flat Rate
Unnamed: 1         0.21
Name: 8, dtype: object

In [68]:
no_flex_cost = data["kwh"].sum() * no_flex[1]
round(no_flex_cost, 2)

1438.34

In [69]:
monthly_flex_cost = round(amount_data["cost"].sum(), 2)
monthly_flex_cost

1421.82

In [70]:
hourly_data = contract.iloc[8:32, 8:10]
hourly_data.rename(columns = {"Unnamed: 8" : "hour", "Unnamed: 9" : "flex"}, inplace = True)
hourly_data = hourly_data.reset_index(drop = True)
hourly_data

hour  flex
0    12AM until 1AM   0.1
1     1AM until 2AM   0.1
2     2AM until 3AM   0.1
3     3AM until 4AM   0.1
4     4AM until 5AM  0.15
5     5AM until 6AM   0.2
6     6AM until 7AM  0.24
7     7AM until 8AM  0.24
8     8AM until 9AM  0.26
9    9AM until 10AM   0.2
10  10AM until 11AM   0.2
11  11AM until 12PM  0.26
12   12PM until 1PM  0.26
13    1PM until 2PM   0.2
14    2PM until 3PM  0.24
15    3PM until 4PM  0.18
16    4PM until 5PM  0.15
17    5PM until 6PM   0.3
18    6PM until 7PM  0.24
19    7PM until 8PM  0.24
20    8PM until 9PM  0.12
21   9PM until 10PM  0.11
22  10PM until 11PM   0.1
23  11PM until 12AM   0.1

In [71]:
data.head()

kwh         day hour   day_name     Month  new_hour                time
0  0.384  2014-03-24  3PM     Monday     March  15:00:00 2014-03-24 15:00:00
1  1.201  2014-08-15  5AM     Friday    August   5:00:00 2014-08-15 05:00:00
2  1.523  2014-03-20  8PM   Thursday     March  20:00:00 2014-03-20 20:00:00
3  0.424  2014-04-23  6PM  Wednesday     April  18:00:00 2014-04-23 18:00:00
4  0.209  2014-12-19  1AM     Friday  December   1:00:00 2014-12-19 01:00:00

In [72]:
data["new_hour2"] = data["new_hour"].astype("datetime64")
data

kwh         day  hour   day_name     Month  new_hour  \
0     0.384  2014-03-24   3PM     Monday     March  15:00:00   
1     1.201  2014-08-15   5AM     Friday    August   5:00:00   
2     1.523  2014-03-20   8PM   Thursday     March  20:00:00   
3     0.424  2014-04-23   6PM  Wednesday     April  18:00:00   
4     0.209  2014-12-19   1AM     Friday  December   1:00:00   
...     ...         ...   ...        ...       ...       ...   
8755  0.084  2014-11-07   1AM     Friday  November   1:00:00   
8756  1.027  2014-05-20   6AM    Tuesday       May   6:00:00   
8757  0.052  2014-04-08   2AM    Tuesday     April   2:00:00   
8758  1.428  2014-01-27   9PM     Monday   January  21:00:00   
8759  0.528  2014-08-25  12PM     Monday    August  12:00:00   

                    time           new_hour2  
0    2014-03-24 15:00:00 2021-03-22 15:00:00  
1    2014-08-15 05:00:00 2021-03-22 05:00:00  
2    2014-03-20 20:00:00 2021-03-22 20:00:00  
3    2014-04-23 18:00:00 2021-03-22 18:00:00  
4    2014-12-19 01:00:00 2021-03-22 01:00:00  
...                  ...                 ...  
8755 2014-11-07 01:00:00 2021-03-22 01:00:00  
8756 2014-05-20 06:00:00 2021-03-22 06:00:00  
8757 2014-04-08 02:00:00 2021-03-22 02:00:00  
8758 2014-01-27 21:00:00 2021-03-22 21:00:00  
8759 2014-08-25 12:00:00 2021-03-22 12:00:00  

[8760 rows x 8 columns]

In [73]:
hourly_cost = data.resample("H", on = "new_hour2")["kwh"].sum()
hourly_cost = hourly_cost.reset_index()
hourly_cost.rename(columns = {"new_hour2" : "hours"}, inplace=True)
hourly_cost

hours      kwh
0  2021-03-22 00:00:00  181.638
1  2021-03-22 01:00:00   94.135
2  2021-03-22 02:00:00   92.192
3  2021-03-22 03:00:00   92.598
4  2021-03-22 04:00:00   93.596
5  2021-03-22 05:00:00  352.153
6  2021-03-22 06:00:00  446.288
7  2021-03-22 07:00:00  399.243
8  2021-03-22 08:00:00  268.043
9  2021-03-22 09:00:00  190.439
10 2021-03-22 10:00:00  183.174
11 2021-03-22 11:00:00  214.968
12 2021-03-22 12:00:00  212.719
13 2021-03-22 13:00:00  184.716
14 2021-03-22 14:00:00  208.166
15 2021-03-22 15:00:00  220.122
16 2021-03-22 16:00:00  218.990
17 2021-03-22 17:00:00  481.243
18 2021-03-22 18:00:00  618.994
19 2021-03-22 19:00:00  734.074
20 2021-03-22 20:00:00  670.219
21 2021-03-22 21:00:00  223.139
22 2021-03-22 22:00:00  210.731
23 2021-03-22 23:00:00  257.671

In [74]:
hourly_cost["hours"] = hourly_cost["hours"].astype("object")

In [75]:
hourly_flex_cost = pd.concat([hourly_data, hourly_cost], axis = 1)
hourly_flex_cost

hour  flex                hours      kwh
0    12AM until 1AM   0.1  2021-03-22 00:00:00  181.638
1     1AM until 2AM   0.1  2021-03-22 01:00:00   94.135
2     2AM until 3AM   0.1  2021-03-22 02:00:00   92.192
3     3AM until 4AM   0.1  2021-03-22 03:00:00   92.598
4     4AM until 5AM  0.15  2021-03-22 04:00:00   93.596
5     5AM until 6AM   0.2  2021-03-22 05:00:00  352.153
6     6AM until 7AM  0.24  2021-03-22 06:00:00  446.288
7     7AM until 8AM  0.24  2021-03-22 07:00:00  399.243
8     8AM until 9AM  0.26  2021-03-22 08:00:00  268.043
9    9AM until 10AM   0.2  2021-03-22 09:00:00  190.439
10  10AM until 11AM   0.2  2021-03-22 10:00:00  183.174
11  11AM until 12PM  0.26  2021-03-22 11:00:00  214.968
12   12PM until 1PM  0.26  2021-03-22 12:00:00  212.719
13    1PM until 2PM   0.2  2021-03-22 13:00:00  184.716
14    2PM until 3PM  0.24  2021-03-22 14:00:00  208.166
15    3PM until 4PM  0.18  2021-03-22 15:00:00  220.122
16    4PM until 5PM  0.15  2021-03-22 16:00:00  218.990
17    5PM until 6PM   0.3  2021-03-22 17:00:00  481.243
18    6PM until 7PM  0.24  2021-03-22 18:00:00  618.994
19    7PM until 8PM  0.24  2021-03-22 19:00:00  734.074
20    8PM until 9PM  0.12  2021-03-22 20:00:00  670.219
21   9PM until 10PM  0.11  2021-03-22 21:00:00  223.139
22  10PM until 11PM   0.1  2021-03-22 22:00:00  210.731
23  11PM until 12AM   0.1  2021-03-22 23:00:00  257.671

In [76]:
hourly_flex_cost["cost"] = hourly_flex_cost["flex"] * hourly_flex_cost["kwh"]
hourly_flex_cost

hour  flex                hours      kwh     cost
0    12AM until 1AM   0.1  2021-03-22 00:00:00  181.638  18.1638
1     1AM until 2AM   0.1  2021-03-22 01:00:00   94.135   9.4135
2     2AM until 3AM   0.1  2021-03-22 02:00:00   92.192   9.2192
3     3AM until 4AM   0.1  2021-03-22 03:00:00   92.598   9.2598
4     4AM until 5AM  0.15  2021-03-22 04:00:00   93.596  14.0394
5     5AM until 6AM   0.2  2021-03-22 05:00:00  352.153  70.4306
6     6AM until 7AM  0.24  2021-03-22 06:00:00  446.288  107.109
7     7AM until 8AM  0.24  2021-03-22 07:00:00  399.243  95.8183
8     8AM until 9AM  0.26  2021-03-22 08:00:00  268.043  69.6912
9    9AM until 10AM   0.2  2021-03-22 09:00:00  190.439  38.0878
10  10AM until 11AM   0.2  2021-03-22 10:00:00  183.174  36.6348
11  11AM until 12PM  0.26  2021-03-22 11:00:00  214.968  55.8917
12   12PM until 1PM  0.26  2021-03-22 12:00:00  212.719  55.3069
13    1PM until 2PM   0.2  2021-03-22 13:00:00  184.716  36.9432
14    2PM until 3PM  0.24  2021-03-22 14:00:00  208.166  49.9598
15    3PM until 4PM  0.18  2021-03-22 15:00:00  220.122   39.622
16    4PM until 5PM  0.15  2021-03-22 16:00:00  218.990  32.8485
17    5PM until 6PM   0.3  2021-03-22 17:00:00  481.243  144.373
18    6PM until 7PM  0.24  2021-03-22 18:00:00  618.994  148.559
19    7PM until 8PM  0.24  2021-03-22 19:00:00  734.074  176.178
20    8PM until 9PM  0.12  2021-03-22 20:00:00  670.219  80.4263
21   9PM until 10PM  0.11  2021-03-22 21:00:00  223.139  24.5453
22  10PM until 11PM   0.1  2021-03-22 22:00:00  210.731  21.0731
23  11PM until 12AM   0.1  2021-03-22 23:00:00  257.671  25.7671

In [77]:
round(hourly_flex_cost["cost"].sum(), 2)

1369.36

### no_flex_cost = 1438.34
### monthly_flex_cost = 1421.82
### hourly_flex_cost = 1369.36


## "Hourly Contract" is the lowest cost